pkg/cmd/pulumi/package_publish.go

```diff
@@ -0,0 +1,241 @@
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/executable"
+	"github.com/spf13/cobra"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+func newPackagePublishCmd() *cobra.Command {
+
+	var err error
+	var path string
+	var tag string
+
+	cmd := &cobra.Command{
+		Use:   "publish-sdk <package> <language>",
+		Args:  cobra.ExactArgs(2),
+		Short: "Publish a package SDK to supported package registries.",
+		Long:  "Publish a package SDK to supported package registries.",
+
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// Validate the language
+
+			lang := args[1]
+
+			switch lang {
+			case "all":
+				fmt.Println("This command is still under construction. " +
+					"Please wait for a future release.")
+			case "nodejs":
+				fmt.Println("Nodejs is supported, publishing")
+			case "python":
+				fmt.Println("This command is still under construction. " +
+					"Please wait for a future release.")
+			case "java":
+				fmt.Println("This command is still under construction. " +
+					"Please wait for a future release.")
+			case "dotnet":
+				fmt.Println("This command is still under construction. " +
+					"Please wait for a future release.")
+			default:
+				return fmt.Errorf(
+					"language %s is not supported. Must be one of: \n"+
+						"- all \n"+
+						"- nodejs \n"+
+						"- python \n"+
+						"- java \n"+
+						"- dotnet ", lang,
+				)
+			}
+
+			return nil
+		},
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			pkg := args[0]
+			lang := args[1]
+
+			err = publish(pkg, lang, path, tag)
+			if err != nil {
+				return fmt.Errorf("error publishing: %s", err.Error())
+			}
+			return nil
+		}),
+	}
+	cmd.PersistentFlags().StringVar(&path, "path", "",
+		`The path to the root of your package.
+	Example: ./sdk/nodejs
+	`)
+	cmd.PersistentFlags().StringVar(&tag, "tag", "",
+		`The tag for your NPM package, such as "latest""`)
+	return cmd
+}
+
+func publish(pkg, lang, path, tag string) error {
+
+	fmt.Printf("You are publishing package %s to the %s registry \n", pkg, lang)
+	switch lang {
+	case "nodejs":
+		err := publishToNPM(pkg, path, tag)
+		if err != nil {
+			return err
+		}
+	case "python":
+		publishToPyPI(pkg)
+	case "java":
+		publishToMaven(pkg)
+	case "dotnet":
+		publishToNuget(pkg)
+	default:
+		// Publish to all registries
+		err := publishToNPM(pkg, path, tag)
+		if err != nil {
+			return err
+		}
+		publishToPyPI(pkg)
+		publishToMaven(pkg)
+		publishToNuget(pkg)
+	}
+
+	return nil
+}
+
+func publishToNPM(pkg, path, tag string) error {
+	// this is likely where we need to go get the sdk info
+	// verify path
+	_, err := os.Stat(path)
+	if errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("could not find path to package directory: %s", path)
```

Comment:
minor note on error messages (again, we're not consistent about this in the codebase but we're trying): instead of doing "could not foo", "error doing bar", these messages that are part of an error object and are meant to be chained should just be "foo" or "bar" so that you don't end up with a chain of "could not foo: could not bar: could not baz: ...".

The following links elaborate on this better than I can in this comment:

- https://github.com/uber-go/guide/blob/master/style.md#error-wrapping
- https://preslav.me/2023/04/14/golang-error-handling-is-a-form-of-storytelling/

Reply:
I understand what you're saying about not continuously wrapping new errors. However, as a user, I find minimalist errors to be unhelpful, and I would very much like to give as much information as I can.

I may disagree with Go conventions here. There is room for returning a meaningful error _and_ avoiding the chaining effect you describe. I'll see what I can do to clean this up.

Reply:
FWIW, the only part I'm asking to omit is the "could not". e.g.:

    return fmt.Errorf("find path to package directory: %s", path)
    // or
    return fmt.Errorf("find package directory: %s", path)


Although in this specific case, it seems that the error should be:

    return fmt.Errorf("package directory does not exist: %s", path)
    // or
    return fmt.Errorf("package directory %q does not exist", path)